src/server/common.h

```diff
@@ -293,9 +289,10 @@ class ExecutionState : protected Cancellation {
   void JoinErrorHandler();
 
  private:
-  // Report error.
   GenericError ReportErrorInternal(GenericError&& err);
 
+  enum class State { RUN, CANCELLED, ERROR };
```

Comment:
We `do not` need these states. You want to figure out if:

1. Operation cancelled 
2. Operation cancelled with an error

Cancelled operation is an error. That's why we have `ReportCancelcedError` which cancels the context with the error `ECANCELED`. Therefore you don't need any of the 3 states here. All you have to do is:

```
  bool IsError() const {
    return ec && !IsCancelled();
  }

  bool IsCancelled() const {
    return ec && err_ == ECANCELED;
  }
```

Reply:
In your state machine:

```
RUN == (ec_ == std::error_code{})
CANCELLED == (ec_ == ECANCELED)
ERROR == (ec_ && !CANCELLED) 
```